include/zephyr/drivers/misc/rpi_pico_pio/rpi_pico_pio.h

```diff
@@ -0,0 +1,207 @@
+ *
+ * @see DT_RPI_PICO_PIO_CLOCK_FREQ_HZ
+ */
+#define DT_INST_RPI_PICO_PIO_CLOCK_FREQ_HZ(inst)						\
+	DT_RPI_PICO_PIO_CLOCK_FREQ_HZ(DT_DRV_INST(inst))
+
+/**
+ * @brief Get a pin number from a pinctrl / group name and index
+ *
+ * Example devicetree fragment(s):
+ *
+ * @code{.dts}
+ *	pinctrl {
+ *		pio_child_default: pio_child_default {
+ *			tx_gpio {
+ *				pinmux = <PIO0_P0>, <PIO0_P2>;
+ *			};
+ *
+ *			rx_gpio {
+ *				pinmux = <PIO0_P1>;
+ *				input-enable;
+ *			};
+ *		};
+ *	};
+ * @endcode
+ *
+ * @code{.dts}
+ *	pio {
+ *		status = "okay";
+ *		c: child {
+ *			pinctrl-0 = <&pio_child_default>;
+ *			pinctrl-names = "default";
+ *		};
+ *	};
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *	DT_RPI_PICO_PIO_PIN_BY_NAME(node, default, 0, tx_gpio, 0) // 0
+ *	DT_RPI_PICO_PIO_PIN_BY_NAME(node, default, 0, tx_gpio, 1) // 2
+ *	DT_RPI_PICO_PIO_PIN_BY_NAME(node, default, 0, rx_gpio, 0) // 1
+ * @endcode
+ *
+ * @param node_id node identifier
+ * @param p_name pinctrl name
+ * @param p_idx pinctrl index
+ * @param g_name group name
+ * @param g_idx group index
+ * @return pin number
+ */
+#define DT_RPI_PICO_PIO_PIN_BY_NAME(node_id, p_name, p_idx, g_name, g_idx)			\
+	RP2_GET_PIN_NUM(DT_PROP_BY_IDX(								\
+		DT_CHILD(DT_PINCTRL_BY_NAME(node_id, p_name, p_idx), g_name), pinmux, g_idx))
+
+/**
+ * @brief Get a pin number from a pinctrl / group name and index
+ *
+ * @param inst instance number
+ * @param p_name pinctrl name
+ * @param p_idx pinctrl index
+ * @param g_name group name
+ * @param g_idx group index
+ * @return pin number
+ *
+ * @see DT_RPI_PICO_PIO_PIN_BY_NAME
+ */
+#define DT_INST_RPI_PICO_PIO_PIN_BY_NAME(inst, p_name, p_idx, g_name, g_idx)			\
+	DT_RPI_PICO_PIO_PIN_BY_NAME(DT_DRV_INST(inst), p_name, p_idx, g_name, g_idx)
+
+/**
+ * @brief Get assigned PIO base address
+ *
+ * @param node_id node identifier
+ * @return base address
+ */
+#define DT_RPI_PICO_PIO_REG_ADDR(node_id)							\
+	DT_REG_ADDR(DT_PARENT(node_id))
+
+/**
+ * @brief Get assigned PIO base address
+ *
+ * @param inst instance number
+ * @return base address
+ *
+ * @see DT_RPI_PICO_PIO_REG_ADDR
+ */
+#define DT_INST_RPI_PICO_PIO_REG_ADDR(inst)							\
+	DT_RPI_PICO_PIO_REG_ADDR(DT_DRV_INST(inst))
+
+/**
+ * @brief PIO shared IRQ handler type
+ *
+ * @param dev Pointer to registered device
+ */
+typedef void (*rpi_pico_pio_irq_t)(const struct device *dev);
+
+/**
+ * @brief PIO shared IRQ configuration data
+ */
+struct rpi_pico_pio_irq_cfg {
+	/* Device node */
+	sys_snode_t node;
+
+	/* IRQ callback */
+	rpi_pico_pio_irq_t irq_func;
+
+	/* Callback parameter */
+	const void *irq_param;
+
+	/* Shared IRQ index */
+	uint8_t irq_idx;
+
+	/* IRQ status */
+	bool enabled;
+};
+
+/**
+ * @brief Register a shared IRQ
+ *
+ * @param dev Pointer to the parent device
+ * @param cfg Pointer to the IRQ configuration
+ */
+void rpi_pico_pio_irq_register(const struct device *dev, struct rpi_pico_pio_irq_cfg *cfg);
```

Comment:
@soburi  

Actually you can have as many devices as you want (can) on a PIO instance. See below for how I'm adding 4 extra uarts (2 per each PIO instance). In practice, you are only limited by the number of SMs / Instruction memory availability as there's no real inter dependency between SMs.

The 'bus' / 'on-bus' thing is mostly for static checking as you are required to be on a PIO parent device. This could be however implemented with a phandle property for each device, but it makes the DTS and checks more confusing.

The intermediate IRQ routing is done for potential load sharing purposes. You could for example overlay PIO IRQ0 to a higher prio and assign it to a child device that requires that higher prio, or you could have one or more child devices that need separate IRQ lines.

```
&pio0 {
	status = "okay";

	pio0_uart0: pio0_uart0 {
		compatible = "raspberrypi,pico-pio-uart";

		pio,interrupts = <0>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio0_uart0_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};

	pio0_uart1: pio0_uart1 {
		compatible = "raspberrypi,pico-pio-uart";

		pio,interrupts = <1>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio0_uart1_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};
};

&pio1 {
	status = "okay";

	pio1_uart0: pio1_uart0 {
		compatible = "raspberrypi,pico-pio-uart";

		pio,interrupts = <0>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio1_uart0_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};

	pio1_uart1: pio1_uart1 {
		compatible = "raspberrypi,pico-pio-uart";

		pio,interrupts = <1>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio1_uart1_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};
};
```